Backport Solidus master Circle CI test scenarios to v3.2

gsmendoza opened the pull request:
## Goal

We want to backport the Solidus master CircleCI config to v3.2
So that we can test various Solidus installation scenarios on v3.2

## Known issues

The `extension:test_app` smoke test from master currently fails in v3.2. 

Given my working Solidus branch is v3.2, when I run the following test script from master in CI:

```
mkdir -p /tmp/dummy_extension
cd /tmp/dummy_extension
bundle init
bundle add rails sqlite3 --skip-install
bundle add solidus --git "file://$(ruby -e"puts File.expand_path ENV['CIRCLE_WORKING_DIRECTORY']")"
export LIB_NAME=set # dummy requireable file
bundle exec rake -rrails -rspree/testing_support/extension_rake -e'Rake::Task["extension:test_app"].invoke'
```

I get the following error:

```
NoMethodError: undefined method `join' for nil:NilClass
https://raw.githubusercontent.com/solidusio/solidus_starter_frontend/v3.2/template.rb:61:in `block in apply'
https://raw.githubusercontent.com/solidusio/solidus_starter_frontend/v3.2/template.rb:5:in `block in apply'
https://raw.githubusercontent.com/solidusio/solidus_starter_frontend/v3.2/template.rb:46:in `apply'
/home/circleci/.rubygems/bundler/gems/solidus-80ccfeda8438/core/lib/generators/solidus/install/install_generator/install_frontend.rb:55:in `install_solidus_starter_frontend'
/home/circleci/.rubygems/bundler/gems/solidus-80ccfeda8438/core/lib/generators/solidus/install/install_generator/install_frontend.rb:19:in `call'
/home/circleci/.rubygems/bundler/gems/solidus-80ccfeda8438/core/lib/generators/solidus/install/install_generator.rb:200:in `install_frontend'
/home/circleci/.rubygems/bundler/gems/solidus-80ccfeda8438/core/lib/spree/testing_support/common_rake.rb:21:in `block (2 levels) in initialize'
/home/circleci/.rubygems/bundler/gems/solidus-80ccfeda8438/core/lib/spree/testing_support/extension_rake.rb:9:in `block (2 levels) in <top (required)>'
(eval):1:in `block in standard_rake_options'
/home/circleci/.rubygems/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/usr/local/bin/bundle:25:in `load'
/usr/local/bin/bundle:25:in `<main>'
Tasks: TOP => common:test_app
```

This is the line from https://raw.githubusercontent.com/solidusio/solidus_starter_frontend/v3.2/template.rb that is failing:

```ruby
repo_dir = Rails.root.join("tmp/solidus_starter_frontend-#{SecureRandom.hex}").tap(&:mkpath).to_s
```

Based on the error, `Rails.root` appears to be returning nil.

See https://app.circleci.com/pipelines/github/nebulab/solidus/564/workflows/7f01bce7-fc69-496c-9b2b-b96aaa10da8a/jobs/6680 for a sample of the failure.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] I have used clear, explanatory commit messages.

The following are not always needed (~cross them out~ if they are not):

- [x] I have added automated tests to cover my changes.
- ~[ ] I have attached screenshots to demo visual changes.~
- ~[ ] I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).~
- ~[ ] I have updated the README to account for my changes.~

elia commented on .circleci/config.yml:
```diff
@@ -176,13 +176,13 @@ jobs:
       - libvips
 
-      - install_solidus: { flags: "--sample=false --frontend=none --authentication=none" }
+      - install_solidus: { flags: "--sample=false --frontend=none --no-with-authentication" }
```
nit:
Should this be `--with-authentication=false`?
I know it works both ways, but my guess is this "spelling" is more common.

gsmendoza replied:
Okay, let's go with --with-authentication=false :+1:

kennyadsl commented on .circleci/config.yml:
```diff
@@ -103,8 +103,103 @@ commands:
+
+            ruby -v >> /tmp/.tool-versions
+            gem search -eq rails >> /tmp/.tool-versions # get the latest rails from rubygems
+            gem search -eq solidus >> /tmp/.tool-versions # get the latest solidus from rubygems
```
Does this work only with the latest solidus? What happen when we will release v3.3?

elia replied:
@kennyadsl I think it will just bust the cache, which is the intended outcome IIRC.

elia commented on .circleci/config.yml:
```diff
@@ -185,17 +185,6 @@ jobs:
-            bundle add solidus --git "file://$(ruby -e"puts File.expand_path ENV['CIRCLE_WORKING_DIRECTORY']")"
-            export LIB_NAME=set # dummy requireable file
-            bundle exec rake -rrails -rspree/testing_support/extension_rake -e'Rake::Task["extension:test_app"].invoke'
-
```
We should keep an eye on this failure as it might be a fluke that can be ignored or a symptom of something broken. Having gone through the same process on master I expect this backport to be equally tricky 🃏

elia commented on .circleci/config.yml:
```diff
@@ -185,7 +191,63 @@ jobs:
+          matrix: { parameters: { rails: ['6.0'], ruby: ['2.6'], database: ['sqlite'], paperclip: [true] } }
+      - test_solidus:
+          name: *name
+          matrix: { parameters: { rails: ['5.2'], ruby: ['2.5'], database: ['sqlite'], paperclip: [true] } }
```
@kennyadsl branch CI rules for v3.2 must be updated after this change to match the new job names 🎗️

elia reviewed (approved):
LGTM, left one optional nit and a concern about the failing test, but it can be added back in a different PR as long as we don't release a patch update in the meantime.

kennyadsl reviewed (commented):
Left a comment + I'd like to investigate the failure on the test_app failure before moving forward, maybe @elia already faced something similar working on master?

Also, I'd remove the screenshot attached to the `[Break down solidus and remove solidus_frontend if frontend=none](https://github.com/solidusio/solidus/pull/4717/commits/31d0c4c7c1a4afc36cdc1cfbcb0852abadaf5445)` commit message. We might lose that reference if we change tool. 

Thanks @gsmendoza 🙏

gsmendoza commented:
> Left a comment + I'd like to investigate the failure on the test_app failure before moving forward, maybe @elia already faced something similar working on master?

@kennyadsl Okay, I'll timebox investigating the issue for a day or two. Will keep you updated.

gsmendoza commented:
@kennyadsl @elia Done:

- Changed authentication flag to --with-authentication=false
- Removed the screenshot attached to the "Break down solidus and remove solidus_frontend if frontend=none" commit

As discussed offline, I'll submit a separate PR for the `extension:test_app` test.
